src/lib/query/pci.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{Iface, NisporError};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(into = "String")]
+#[serde(try_from = "String")]
+pub struct PciAddress {
+    pub domain: u32,
+    pub bus: u8,
+    pub slot: u8,
+    pub function: u8,
+}
+
+impl std::fmt::Display for PciAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}:{:02x}:{:02x}.{}",
+            if self.domain > u16::MAX.into() {
+                format!("{:08x}", self.domain)
+            } else {
+                format!("{:04x}", self.domain)
+            },
+            self.bus,
+            self.slot,
+            self.function
+        )
```

Comment:
Better
```
       write!(
            f,
            if self.domain > u16::MAX.into() {
                "{:08x}:{:02x}:{:02x}.{}"
            } else {
                 "{:04x}:{:02x}:{:02x}.{}"
            },
            self.domain,
            self.bus,
            self.slot,
            self.function
        )
```
It avoids a string allocation.

Reply:
```
error: format argument must be a string literal
  --> src/lib/query/pci.rs:24:13
   |
24 | /             if self.domain > u16::MAX.into() {
25 | |                 "{:08x}:{:02x}:{:02x}.{}"
26 | |             } else {
27 | |                 "{:04x}:{:02x}:{:02x}.{}"
28 | |             },
   | |_____________^
   |
help: you might be missing a string literal to format with
   |
24 |             "{} {} {} {} {}", if self.domain > u16::MAX.into() {
   |             +++++++++++++++++

error: could not compile `nispor` (lib) due to 1 previous error
```

Reply:
You're right.